shared-bindings/alarm/__init__.c

```diff
@@ -0,0 +1,233 @@
+        common_hal_alarm_light_sleep_until_alarms(n_args, args);
+    }
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(alarm_light_sleep_until_alarms_obj, 1, MP_OBJ_FUN_ARGS_MAX, alarm_light_sleep_until_alarms);
+
+//| def exit_and_deep_sleep_until_alarms(*alarms: Alarm) -> None:
+//|     """Exit the program and go into a deep sleep, until awakened by one of the alarms.
+//|     This function does not return.
+//|
+//|     When awakened, the microcontroller will restart and will run ``boot.py`` and ``code.py``
+//|     from the beginning.
+//|
+//|     After restart, an alarm *equivalent* to the one that caused the wake-up
+//|     will be available as `alarm.wake_alarm`.
+//|     Its type and/or attributes may not correspond exactly to the original alarm.
+//|     For time-base alarms, currently, an `alarm.time.TimeAlarm()` is created.
+//|
+//|     If no alarms are specified, the microcontroller will deep sleep until reset.
+//|
+//|     **If CircuitPython is connected to a host computer, `alarm.exit_and_deep_sleep_until_alarms()`
+//|     then the connection will be maintained, and the system will not go into deep sleep.**
+//|     This allows the user to interrupt an existing program with ctrl-C,
+//|     and to edit the files in CIRCUITPY, which would not be possible in true deep sleep.
+//|     Thus, to use deep sleep and save significant power, you will need to disconnect from the host.
+//|
+//|     Here is skeletal example that deep-sleeps and restarts every 60 seconds:
+//|
+//|     .. code-block:: python
+//|
+//|         import alarm
+//|         import time
+//|
+//|         print("Waking up")
+//|
+//|         # Set an alarm for 60 seconds from now.
+//|         time_alarm = alarm.time.TimeAlarm(monotonic_time=time.monotonic() + 60)
+//|
+//|         # Deep sleep until the alarm goes off. Then restart the program.
+//|         alarm.exit_and_deep_sleep_until_alarms(time_alarm)
+//|     """
+//|     ...
+//|
+STATIC mp_obj_t alarm_exit_and_deep_sleep_until_alarms(size_t n_args, const mp_obj_t *args) {
+    validate_objs_are_alarms(n_args, args);
+
+    // Shut down WiFi, etc.
+    common_hal_alarm_prepare_for_deep_sleep();
+
+    // See if we are connected to a host.
+    // Make sure we have been awake long enough for USB to connect (enumeration delay).
+    int64_t connecting_delay_msec = CIRCUITPY_USB_CONNECTED_SLEEP_DELAY * 1024 - supervisor_ticks_ms64();
+    if (connecting_delay_msec > 0) {
+        common_hal_time_delay_ms(connecting_delay_msec * 1000 / 1024);
+    }
+
+    if (supervisor_workflow_active()) {
+        // Simulate deep sleep by waiting for an alarm and then restarting when done.
+        common_hal_alarm_wait_until_alarms(n_args, args);
+        reload_requested = true;
+        supervisor_set_run_reason(RUN_REASON_STARTUP);
+        mp_raise_reload_exception();
+    } else {
+        common_hal_alarm_exit_and_deep_sleep_until_alarms(n_args, args);
+        // Does not return.
+    }
```

Comment:
This logic should be moved to main.c where other workflow code is.

Reply:
I did not move all of this code because it is very similar to the code on lines 98-109, for the light sleep routine, which would not go back to `main.c` because it does not exit. I _could_ move the logic on lines 167-169, probably by creating another deep-sleep exception (or passing a different value). I'm not sure I want to do that because it separates the logic into two disparate pieces. Could you explain more about why you want to split it?

Reply:
I thought about this some more this morning, and did move the restart to main.c.

This signaling is done by internal-only exceptions `_DeepSleepRequest` and `_DeepSleepRestart`. An alternative would be to pass the `PYEXEC flags as the `SysExit` exception argument. However, the MicroPython base code currently doesn't handle the argument at all, and it's a substantial drift away the MicroPython base to do that.